src/join/core.jl

```diff
@@ -0,0 +1,534 @@
+
+    left_ixs = Int[]
+    right_ixs = Int[]
+
+    sizehint!(left_ixs, right_len)
+    sizehint!(right_ixs, right_len)
+
+    compute_join_indices!(groups, ngroups, starts, rperm)
+
+    n = 0
+    @inbounds for (idx_l, val_l) in enumerate(left)
+        if ismissing(val_l)
+            group_id = dict[end]
+        elseif minv <= val_l <= maxv
+            group_id = dict[Int(val_l) + offset]
+        else
+            group_id = 0
+        end
+
+        if group_id > 0
+            ref_stop = starts[group_id + 1]
+            l = ref_stop - starts[group_id]
+            newn = n + l
+            resize!(left_ixs, newn)
+            for i in n+1:n+l
+                left_ixs[i] = idx_l
+            end
+            resize!(right_ixs, newn)
+            for i in 1:l
+                right_ixs[n + i] = rperm[ref_stop - i + 1]
+            end
+            n = newn
+        end
+    end
+
+    return left_ixs, right_ixs
+end
+
+function find_inner_rows(joiner::DataFrameJoiner)
+    right_len = length(joiner.dfr_on[!, 1])
+    left_len = length(joiner.dfl_on[!, 1])
+    right_shorter = right_len <= left_len
+
+    left_cols = collect(eachcol(joiner.dfl_on))
+    right_cols = collect(eachcol(joiner.dfr_on))
+
+    # if column of a longer table supports DataAPI.refpool and DataAPI.invrefpool
+    # remap matching left and right columns to use refs
+    if right_shorter
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+
+            lc_refs = DataAPI.refarray(lc)
+            lc_refpool = DataAPI.refpool(lc)
+            lc_invrefpool = DataAPI.invrefpool(lc)
+            if check_mapping_allowed(rc, lc_refs, lc_refpool, lc_invrefpool)
+                right_cols[i] = map2refs(rc, lc_invrefpool)
+                left_cols[i] = lc_refs
+            end
+        end
+    else
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+
+            rc_refs = DataAPI.refarray(rc)
+            rc_refpool = DataAPI.refpool(rc)
+            rc_invrefpool = DataAPI.invrefpool(rc)
+            if check_mapping_allowed(lc, rc_refs, rc_refpool, rc_invrefpool)
+                right_cols[i] = rc_refs
+                left_cols[i] = map2refs(lc, rc_invrefpool)
+            end
+        end
+    end
+
+    # this is a workaround for https://github.com/JuliaData/CategoricalArrays.jl/issues/319
+    # this path will be triggered only in rare cases when the refpool code above
+    # fails to convert CategoricalArray into refpool
+    disallow_sorted = false
+
+    for (lc, rc) in zip(left_cols, right_cols)
+        @assert length(lc) == left_len
+        @assert length(rc) == right_len
+        lct = typeof(lc)
+        lcat = nameof(lct) === :CategoricalArray && nameof(parentmodule(lct)) === :CategoricalArrays
+        rct = typeof(rc)
+        rcat = nameof(rct) === :CategoricalArray && nameof(parentmodule(rct)) === :CategoricalArrays
```

Comment:
@nalimilan - we really need to resolve https://github.com/JuliaData/CategoricalArrays.jl/issues/319, as the code I have marked is incorrect. It does not correctly handle the following cases:
* view of `CategoricalVector` (we do not catch it)
* a `Vector` holding `CategoricalValue` entries (potentially mixed with non-categorical entries of the same unwrapped type)

I was thinking if it is possible to do it fully correctly without https://github.com/JuliaData/CategoricalArrays.jl/issues/319 but I do not have a good solution (in particular the second case when `Vector` can have `eltype` `Any`). I can fix this code to cover containers for which `eltype` is `CategoricalValue` (potentially mixed with `missing`) but it is not fully correct either and will be even more messy.

How do you think we should move forward? I think we need to make this decision (maybe not releasing everything but to have a decision and have the code here reflect this decision) before merging this PR.

Reply:
Yes we should implement that. In the meantime, I don't think we should care about `Vector`s holding `CategoricalValue`, these are quite hard to create and terribly inefficient and inconvenient anyway. Checking the `eltype` should be enough.